nomos-services/mix/src/backends/libp2p.rs

```diff
@@ -0,0 +1,224 @@
+use futures::StreamExt;
+use libp2p::{
+    core::transport::ListenerId,
+    identity::{ed25519, Keypair},
+    swarm::SwarmEvent,
+    Multiaddr, PeerId, Swarm, SwarmBuilder, TransportError,
+};
+use nomos_libp2p::{secret_key_serde, DialError, DialOpts, Protocol};
+use overwatch_rs::{overwatch::handle::OverwatchHandle, services::state::NoState};
+use rand::seq::IteratorRandom;
+use serde::{Deserialize, Serialize};
+use tokio::{
+    sync::{broadcast, mpsc},
+    task::JoinHandle,
+};
+
+use super::NetworkBackend;
+
+pub struct Libp2pNetworkBackend {
+    #[allow(dead_code)]
+    task: JoinHandle<()>,
+    msgs_tx: mpsc::Sender<Libp2pNetworkBackendMessage>,
+    events_tx: broadcast::Sender<Libp2pNetworkBackendEvent>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Libp2pNetworkBackendSettings {
+    pub listening_address: Multiaddr,
+    #[serde(with = "secret_key_serde", default = "ed25519::SecretKey::generate")]
+    pub node_key: ed25519::SecretKey,
+    pub membership: Vec<Multiaddr>,
+    pub peering_degree: usize,
+    pub num_mix_layers: usize,
+}
+
+#[derive(Debug)]
+pub enum Libp2pNetworkBackendMessage {
+    Mix(Vec<u8>),
+}
+
+#[derive(Debug)]
+pub enum Libp2pNetworkBackendEventKind {
+    FullyMixedMessage,
+}
+
+#[derive(Debug, Clone)]
+pub enum Libp2pNetworkBackendEvent {
+    FullyMixedMessage(Vec<u8>),
+}
+
+const CHANNEL_SIZE: usize = 64;
+
+#[async_trait]
+impl NetworkBackend for Libp2pNetworkBackend {
+    type Settings = Libp2pNetworkBackendSettings;
+    type State = NoState<Self::Settings>;
+    type Message = Libp2pNetworkBackendMessage;
+    type EventKind = Libp2pNetworkBackendEventKind;
+    type NetworkEvent = Libp2pNetworkBackendEvent;
+
+    fn new(config: Self::Settings, overwatch_handle: OverwatchHandle) -> Self {
+        let (msgs_tx, msgs_rx) = mpsc::channel(CHANNEL_SIZE);
+        let (events_tx, _) = broadcast::channel(CHANNEL_SIZE);
+
+        let keypair = Keypair::from(ed25519::Keypair::from(config.node_key.clone()));
+        let local_peer_id = keypair.public().to_peer_id();
+        let mut swarm = MixSwarm::new(keypair, config.num_mix_layers, msgs_rx, events_tx.clone());
+
+        swarm
+            .listen_on(config.listening_address)
+            .unwrap_or_else(|e| {
+                panic!("Failed to listen on Mix network: {e:?}");
+            });
+
+        // Randomly select peering_degree number of peers, and dial to them
+        config
+            .membership
+            .iter()
+            .filter(|addr| match extract_peer_id(addr) {
+                Some(peer_id) => peer_id != local_peer_id,
+                None => false,
+            })
+            .choose_multiple(&mut rand::thread_rng(), config.peering_degree)
+            .iter()
+            .cloned()
+            .for_each(|addr| {
+                if let Err(e) = swarm.dial(addr.clone()) {
+                    tracing::error!("failed to dial to {:?}: {:?}", addr, e);
+                }
+            });
```

Comment:
Eventually this should probably be part of the mixnet behavior as well. 
We will probably handle it internally in the da behaviors too. But for now this should do.

Reply:
Maybe we can add a `todo` here.

Reply:
Good point 3a6de0887c6f116db0d08377cf0af83f7d673562